File: src/main/java/com/cartrecipes/service/recipe/RecipeService.java
package com.cartrecipes.service.recipe;

import com.cartrecipes.model.Recipe;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collection;

@Service
@RequiredArgsConstructor
@Transactional
public class RecipeService {
    private final RecipeRepository recipeRepository;

    @Transactional(readOnly = true)
    public Page<Recipe> findAll(Pageable pageable) {
        return recipeRepository.findAll(pageable);
    }

    @Transactional(readOnly = true)
    public Collection<Recipe> findByIds(Collection<Long> ids) {
        return recipeRepository.findAllById(ids);
    }

    @Transactional(readOnly = true)
    public Recipe findById(Long id) {
        return recipeRepository.findWithProductsById(id)
                .orElseThrow(() -> new IllegalArgumentException("Recipe not found with id: " + id));
    }
}

File: src/main/kotlin/com/cartrecipes/service/recipe/RecipeService.kt
package com.cartrecipes.service.recipe

import com.cartrecipes.exception.NotFoundException
import com.cartrecipes.model.Recipe
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class RecipeService internal constructor (
    private val recipeRepository: RecipeRepository
) {
    @Transactional(readOnly = true)
    fun findAll(pageable: Pageable): Page<Recipe> {
        return recipeRepository.findAll(pageable)
    }

    @Transactional(readOnly = true)
    fun findByIds(ids: Collection<Long>): Collection<Recipe> {
        return recipeRepository.findAllById(ids)
    }

    @Transactional(readOnly = true)
    fun findWithProductsById(id: Long): Recipe {
        return recipeRepository.findWithProductsById(id) ?: throw NotFoundException("Recipe not found with id: $id")
    }

    @Transactional(readOnly = true)
    fun findById(id: Long): Recipe {
        return recipeRepository.findById(id).orElseThrow { NotFoundException("Recipe not found with id: $id") }
    }
}